MoneyDataType/AmountConverter.cs

```diff
@@ -1,100 +1,98 @@
-        private const string Dec = "dec";
+        var value = default(decimal);
+        ICurrency currency = null;
+        string nativeName = null;
+        string englishName = null;
+        string symbol = null;
+        string iso = null;
+        int? dec = null;
 
-        public override Amount Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        while (reader.Read() && reader.TokenType == JsonTokenType.PropertyName)
         {
-            var val = default(decimal);
-            ICurrency currency = null;
-            string nativename = null;
-            string englishname = null;
-            string symbol = null;
-            string iso = null;
-            int? dec = null;
+            var propertyName = reader.GetString();
+            if (!reader.Read()) continue;
 
-            while (reader.Read())
+            switch (propertyName)
             {
-                if (reader.TokenType != JsonTokenType.PropertyName) break;
-
-                var propertyName = reader.GetString();
-                if (!reader.Read()) continue;
+                case ValueName:
+                    value = reader.GetDecimal();
+                    break;
+                case CurrencyName:
+                    currency = Currency.FromIsoCode(reader.GetString());
+                    break;
+                case Name: // Kept for backwards compatibility
+                case NativeName:
+                    nativeName = reader.GetString();
+                    break;
+                case EnglishName:
+                    englishName = reader.GetString();
+                    break;
+                case Symbol:
+                    symbol = reader.GetString();
+                    break;
+                case Iso:
+                    iso = reader.GetString();
+                    break;
+                case Dec when reader.TryGetInt32(out var i):
+                    dec = i;
+                    break;
+                default:
+                    throw new InvalidOperationException($"Unknown property name \"{propertyName}\".");
+            }
+        }
 
-                switch (propertyName)
-                {
-                    case ValueName:
-                        val = reader.GetDecimal();
-                        break;
-                    case CurrencyName:
-                        currency = Currency.FromISOCode(reader.GetString());
-                        break;
+        if (!Currency.IsKnownCurrency(currency?.CurrencyIsoCode ?? string.Empty))
+        {
+            currency = new Currency(nativeName, englishName, symbol, iso, dec.GetValueOrDefault(2));
+        }
 
-                    // Kept for backwards compatibility
-                    case Name:
-                        nativename = reader.GetString();
-                        break;
+        if (currency is null) throw new InvalidOperationException("Invalid amount format. Must include a currency");
```

Comment:
I can't move it above the `if` block, because that block sets `currency`. Doing so will [fail the test](https://github.com/OrchardCMS/OrchardCore.Commerce/runs/6193741576?check_suite_focus=true).